Add upwind penalty boundary correction for scalar wave

nilsdeppe opened the pull request:
## Proposed changes

- Have the damping parameter `gamma2` be a temporary tag in the `TimeDerivative` struct. This is needed for boundary corrections. In the future I'd like the `TimeDerivative` struct to actually compute the damping parameter as opposed to having it be a compute tag, but until that happens we need this design. I don't want to allow slicing of arbitrary data from the DataBox because that encourages compute tag use, which we are trying to reduce.
- Add upwind penalty boundary correction. This is basically a refactor of the existing one, changing it to conform to the new interface (as in, the math doesn't change). However, since I don't want to break scalar wave evolutions in develop (and nobody wants to review a 25k line PR), I'm copy-pasting (with some refactor) the boundary correction, and will delete the old one once the evolution code is completely ported. The missing major piece is boundary corrections for the switch to happen.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Make constraint damping a temp variable in scalar wave

nilsdeppe committed:
Add upwind penalty for scalar wave

fmahebert commented on src/Evolution/Systems/ScalarWave/TimeDerivative.cpp:
```diff
@@ -17,12 +17,16 @@ void TimeDerivative<Dim>::apply(
     const gsl::not_null<tnsr::i<DataVector, Dim, Frame::Inertial>*> dt_phi,
     const gsl::not_null<Scalar<DataVector>*> dt_psi,
 
+    const gsl::not_null<Scalar<DataVector>*> result_gamma2,
```
For my education, can you explain this change (i.e., the first commit of this PR) in a github comment?

The caller already knows gamma2, or it wouldn't be able to pass it into the function. What does returning gamma2 again by not-null accomplish?

nilsdeppe replied:
In order to get the constraint damping parameter on the boundary we need it to be a temporary tag. I'm trying to avoid allowing people to slice everything in the DataBox to the boundary because all you need are the principal part terms of the evolution equations and the evolved (maybe prims too) variables. Ideally I'd like the time derivatives to take the gamma2 function as an argument as well as the coordinates, and then compute it internally. The downside of course is that it would get re-computed every time step, even when that's not needed.

fmahebert replied:
Okay, that fills in the gap I was missing -- thank you!

nilsdeppe replied:
Maybe I should rename the parameter for now? Something like `gamma2_for_projection_to_interface`? Lengthy, but documents why it exists :P

fmahebert replied:
That is pretty long... but it only appears in a few places, so it wouldn't be too bad. Alternatively, a short comment (maybe not even a doxygen comment, but a simple `//` comment around line 28 below?) Up to you, I think either would be helpful :)

nilsdeppe commented:
Rebased and squashed in a comment explaining the argument :)

moxcodes commented on src/Evolution/Systems/ScalarWave/BoundaryCorrections/BoundaryCorrection.hpp:
```diff
@@ -0,0 +1,44 @@
+
+/*!
+ * \brief The base class used to create boundary corrections from input files
+ * and store them in the global cache.
```
I'm not sure whether the statements about the global cache makes sense here -- the factory-creatable interface is pretty agnostic about where the pointer ends up.

nilsdeppe replied:
I would like some indication as to why these aren't fully compile time like what we currently do. What about:
```cpp
 /*!                                                                                                                                                                                                                                                                              
  * \brief The base class used to make boundary corrections factory createable so                                                                                                                                                                                                 
  * they can be specified in the input file.                                                                                                                                                                                                                                      
  */
```

moxcodes replied:
That sounds better to me. :+1:

moxcodes commented on src/Evolution/Systems/ScalarWave/BoundaryCorrections/BoundaryCorrection.hpp:
```diff
@@ -0,0 +1,44 @@
+ * and store them in the global cache.
+ */
+template <size_t Dim>
+class BoundaryCorrection : public PUP::able {
```
Is the interface for `dg_package_data` or `dg_boundary_terms` general? Can they be added as additional virtual functions -- that would seem to help a lot with the interface design (implementations wouldn't have to try to loop over possibilities and `dynamic_cast` before using the `BoundaryCorrection`).

nilsdeppe replied:
No, they depend on the boundary correction. The easiest example is that a Rusanov solver only requires the max absolute char speed, while HLL requires both the min and max speeds. Some solvers require characteristic speeds, etc.

moxcodes commented on src/Evolution/Systems/ScalarWave/BoundaryCorrections/Factory.hpp:
```diff
@@ -0,0 +1,7 @@
+
+#pragma once
+
+#include "Evolution/Systems/ScalarWave/BoundaryCorrections/BoundaryCorrection.hpp"
```
Does this serve only to aggregate the headers? I remember discussion against that practice in the past, but perhaps there is a good reason to be in favor of it now?

nilsdeppe replied:
Yes, I don't remember if we every discussed not using this design. I prefer it because I've sometimes changed a header file for 1 domain creater and had something like 150 targets rebuild, all because of the coupling having the includes be in the base class header file.

moxcodes commented on src/Evolution/Systems/ScalarWave/BoundaryCorrections/RegisterDerived.cpp:
```diff
@@ -0,0 +1,13 @@
+
+namespace ScalarWave::BoundaryCorrections {
+void register_derived_with_charm() noexcept {
+  Parallel::register_derived_classes_with_charm<BoundaryCorrection<1>>();
```
Why prefer registering all dimensions? Doesn't any particular executable usually just want the `BoundaryCorrection` associated with the dimension of the system?

nilsdeppe replied:
No, but as far as I know it doesn't do any harm and means there's one fewer template parameter needed.

moxcodes commented on src/Evolution/Systems/ScalarWave/BoundaryCorrections/UpwindPenalty.hpp:
```diff
@@ -0,0 +1,279 @@
+ * by \f$v_{\hat{\alpha}}\f$, and implicitly sum over reapeated indices.
+ * \f$T_{\alpha\hat{\alpha}}\f$ transforms characteristic fields to evolved
+ * fields, while \f$\Lambda_{\hat{\alpha}\hat{\beta}}^-\f$ is a diagonal matrix
+ * with only the negative characteristic speeds. The int and ext superscripts
```
`with only the negative characteristic speeds`->`with only negative characteristic speeds, and has zeros on the diagonal for all other entries` (just to be completely unambiguous?)

moxcodes commented on src/Evolution/Systems/ScalarWave/BoundaryCorrections/UpwindPenalty.hpp:
```diff
@@ -0,0 +1,279 @@
+ *        \end{array}\right.
+ * \f}
+ *
+ * Note that we have assumed \f$n_i^{\mathrm{ext}}\f$ points in the same
```
`\f$n\f$` hasn't been defined in this documentation. Most will probably figure it out, but it would be helpful to spell it out.

moxcodes commented on src/Evolution/Systems/ScalarWave/BoundaryCorrections/UpwindPenalty.hpp:
```diff
@@ -0,0 +1,279 @@
+ * have:
+ *
+ * \f{align}{
+ *   D_{\Psi} &= \bar{\lambda}_{v^{\Psi}}^{\mathrm{ext}}
```
oh geez is the only difference between this formula and the previous one that half the lambdas are barred instead of tilded? That's... pretty hard to read. Perhaps a different adornment?

moxcodes commented on src/Evolution/Systems/ScalarWave/BoundaryCorrections/UpwindPenalty.cpp:
```diff
@@ -0,0 +1,224 @@
+  // Use v_psi allocation as n^i Phi_i
+  dot_product(packaged_char_speed_v_psi, normal_covector, phi);
+  get(*packaged_char_speed_gamma2_v_psi) = get(constraint_gamma2) * get(psi);
+  {
```
in code comment state that the scope is to limit the region in which `packaged_char_speed_v_psi` and `packaged_char_speed_gamma2_v_psi` are borrowed as buffers.
Also, I think this would be less confusing if the temporary use as a buffer was changed to:
```
auto& normal_dot_phi = get(*packaged_char_speed_v_psi);
dot_product(make_not_null(&normal_dot_phi), normal_covector, phi));
```
(i.e. just consistently treat manipulations associated with a particular temporary identity for the buffer to avoid confusion associated with the variable names). If you choose to do this, maybe consider similar for `packaged_char_speed_gamma2_v_psi`.

fmahebert replied:
I agree with Jordan's comment and it doesn't look addressed to me. I'd like to see
- the comment on line 79 move one line down (to below line 80), and say "In this scope, use v_psi allocation as n^i Phi_i"
- then reorder the `dot_product` and the alias creation lines below, so the scope starts with creating the `normal_dot_phi` alias.

nilsdeppe replied:
point 2: I think the reorder is a bad decision because then `normal_dot_phi` can't be const _and_ I'd need to do a `make_not_null`

fmahebert replied:
okay, that's a fair counter-argument 👍

moxcodes commented on src/Evolution/Systems/ScalarWave/BoundaryCorrections/UpwindPenalty.hpp:
```diff
@@ -0,0 +1,279 @@
+ *   D_{\Psi} &= \bar{\lambda}_{v^{\Psi}}^{\mathrm{ext}}
+ *                v^{\mathrm{ext},\Psi}
+ *                - \tilde{\lambda}_{v^{\Psi}}^{\mathrm{int}}
+ *                v^{\mathrm{int},g}, \\
```
this has a `g` in the superscript, which I think must be a mistake

nilsdeppe replied:
Yea, that's a typo left over from the GH upwind penalty... Thanks :)

moxcodes commented on src/Evolution/Systems/ScalarWave/BoundaryCorrections/UpwindPenalty.hpp:
```diff
@@ -0,0 +1,279 @@
+ *
+ * with characteristic fields
+ *
+ * \f{align}{
```
might be nice to note the storage convention for the char speeds -- it makes sense, but it's not immediately obvious.

nilsdeppe replied:
As in, which component maps to which speed?

moxcodes replied:
Yeah, exactly.

moxcodes commented on src/Evolution/Systems/ScalarWave/BoundaryCorrections/UpwindPenalty.hpp:
```diff
@@ -0,0 +1,279 @@
+    using type = Scalar<DataVector>;
+  };
+  struct CharSpeedsTensor : db::SimpleTag {
+    using type = tnsr::i<DataVector, 3, Frame::Inertial>;
```
Why use tensor rather than a `std::array`? As far as I understand, the char speed entries don't really have a direct connection with spacetime directions.

nilsdeppe replied:
Yes, but I can't easily store a `std::array` because all these tags are stuck into a `Variables`. I unfortunately am largely stuck with this "use a Tensor as an array" business

moxcodes commented on src/Evolution/Systems/ScalarWave/BoundaryCorrections/UpwindPenalty.hpp:
```diff
@@ -0,0 +1,279 @@
+ * Note that we have assumed \f$n_i^{\mathrm{ext}}\f$ points in the same
+ * direction as \f$n_i^{\mathrm{int}}\f$. If \f$n_i^{\mathrm{ext}}\f$ points in
+ * the opposite direction the external speeds have their sign flipped and the
+ * \f$\pm\f$ fields and their speeds reverse roles. Specifically, in the code we
```
I'm not sure I understand the `speeds reverse roles` with the formula as written -- it appears from the current form that the v^+ and v^- appear in the same way as the previous equation.

nilsdeppe replied:
I'm adding:
```cpp
  * \f$\pm\f$ fields and their speeds reverse roles (i.e. the                                                                                                                                                                                                                     
  * \f$v^{\mathrm{ext},+}\f$ field is now flowing into the element, while                                                                                                                                                                                                         
  * \f$v^{\mathrm{ext},-}\f$ flows out). Specifically, in the code we have:
```
Does that clarify things? I think the role reversal only matters in the `Phi_i` term because they aren't being added, and there it cancels out with the minus sign that the `n_i^ext` brings in

moxcodes commented on src/Evolution/Systems/ScalarWave/BoundaryCorrections/UpwindPenalty.cpp:
```diff
@@ -0,0 +1,224 @@
+  DataVector& weighted_lambda_psi_int = get(get<::Tags::TempScalar<0>>(buffer));
+  weighted_lambda_psi_int = step_function(-char_speeds_int[0]);
+  DataVector& weighted_lambda_psi_ext = get(get<::Tags::TempScalar<1>>(buffer));
+  weighted_lambda_psi_ext = -step_function(char_speeds_ext[0]);
```
From the documentation, I anticipated some check regarding the `\f$n\f$` vectors, but here it looks like the only version implemented is the one in which `n_i^{ext}` points in the opposite direction to `n_i^{int}`? That makes sense to me, logically, but in the presentation of the documentation it would be helpful to point out that the implementation makes this assumption.

moxcodes commented on src/Evolution/Systems/ScalarWave/BoundaryCorrections/UpwindPenalty.cpp:
```diff
@@ -0,0 +1,224 @@
+  // of the current element. Since lambda_psi from the neighbor is computing
+  // with the normal vector pointing into the current element in the code,
+  // we need to swap the sign of lambda_psi^{ext}. Theta is the heaviside step
+  // function with Theta(0) = 0.
```
`with Theta(0) = 0` is inconsistent with the documentation (which gives the value at \le 0.0). I think our version of `step_function` agrees with the dox and not `Theta(0) = 0`.

nilsdeppe replied:
It doesn't matter in this case since we always have `Theta(Lambda) * Lambda`. Having said that, I'm not sure what you mean by "the documentation" and "the dox". The one on `UpwindPenalty` or `step_function`?

moxcodes replied:
I mean the `UpwindPenalty` dox. There it is a `\le`, where here you imply strict inequality. Perhaps just omit the part of this code comment that states the behavior at 0?

moxcodes commented on src/Evolution/Systems/ScalarWave/BoundaryCorrections/UpwindPenalty.cpp:
```diff
@@ -0,0 +1,224 @@
+  for (size_t d = 0; d < Dim; ++d) {
+    // Overall minus sign on ext because of normal vector is opposite direction.
+    phi_boundary_correction->get(d) =
+        -0.5 * (weighted_lambda_minus_ext *
```
should this extra sign also appear in the final formula in the documentation where it notes the sign flips?

nilsdeppe replied:
Yep!

nilsdeppe replied:
Wait, this is in agreement with the documentation as far as I can tell.

Edit: maybe the code comment should be remove and the minus sign brought in?

moxcodes commented on tests/Unit/Evolution/Systems/ScalarWave/BoundaryCorrections/UpwindPenalty.py:
```diff
@@ -0,0 +1,163 @@
+    ext_char_speed_gamma2_v_psi, ext_char_speeds, use_strong_form):
+    result = int_char_speed_v_psi * 0.
+    # Add v^+ terms
+    if ext_char_speeds[1] > 0.:
```
probably best to be consistent with how the step function in the code and dox works at 0. Is the char speed of 0.0 case tested?

nilsdeppe replied:
At least for some of the char speeds, yes. The thing is that it's always `Theta(Lambda) * Lambda` so it is zero at zero in this case. I'll change it so it's consistent with the C++

nilsdeppe commented:
I've pushed a fixup with some "I hope this resolves the comment" changes, so please let me know if anything isn't correctly addressed!

moxcodes commented:
Fixup looks good to me -- All of my comments are satisfied.
Go ahead and squash.

fmahebert commented on src/Evolution/Systems/ScalarWave/BoundaryCorrections/UpwindPenalty.hpp:
```diff
@@ -0,0 +1,288 @@
+ * \f$D_\alpha\f$. The general form is given by:
+ *
+ * \f{align*}{
+ *   \label{eq:pnpm upwind boundary term characteristics}
```
Why the label? Are you planning to refer to this equation from elsewhere in the dox?

fmahebert commented on src/Evolution/Systems/ScalarWave/BoundaryCorrections/UpwindPenalty.hpp:
```diff
@@ -0,0 +1,288 @@
+ *   v^{\mathrm{ext}}_{\hat{\alpha}}
+ *   -T_{\beta\hat{\beta}}^{\mathrm{int}}
+ *   \Lambda^{\mathrm{int},-}_{\hat{\beta}\hat{\alpha}}
+ *   v^{\mathrm{int}}_{\hat{\alpha}},
```
You are starting a new sentence after this equation, so change the final comma -> period.

fmahebert commented on src/Evolution/Systems/ScalarWave/BoundaryCorrections/UpwindPenalty.hpp:
```diff
@@ -0,0 +1,288 @@
+ * We denote the evolved fields by \f$u_{\alpha}\f$, the characteristic fields
+ * by \f$v_{\hat{\alpha}}\f$, and implicitly sum over reapeated indices.
+ * \f$T_{\alpha\hat{\alpha}}\f$ transforms characteristic fields to evolved
+ * fields, while \f$\Lambda_{\hat{\alpha}\hat{\beta}}^-\f$ is a diagonal matrix
```
[minor nitpick] In lines 35-51, you often use the opposite subscript alpha/beta when talking about the terms vs in the displayed equation. Can you do some alpha <-> beta swaps either in the displayed equation or in the sentences? It's just notation, but might as well make it easier to read...

nilsdeppe replied:
Sorry, you would like me to use `u_beta` and `v_alpha` so it matches the equation?

fmahebert replied:
I'm suggesting either (note: i'll write (a,b) instead of (alpha,beta) for simplicity here),
1) you match the text to the equation by changing D_a -> D_b on line 35, then T_aa -> T_bb and L_ab -> L_ba in this paragraph, OR
2) match the equation to the text by changing D_b -> D_a, T_bb -> T_aa, L_ba -> L_ab, v_a -> v_b in the equation (and then also use u_b and v_b in this paragraph)

I guess it depends if you prefer to have the alpha subscript on D or on (u and v).

kidder commented:
@geoffrey4444 do you want to look at this?

fmahebert commented on src/Evolution/Systems/ScalarWave/BoundaryCorrections/UpwindPenalty.hpp:
```diff
@@ -0,0 +1,288 @@
+ * unit normal covector to the interface. We have also defined
+ *
+ * \f{align}{
+ *   \tilde{\lambda}_{\hat{\alpha}} =
```
In lines 40-50ish, you use `\Lambda^{-}` for a matrix with only negative char speeds. It would be notationally simpler to use `\lambda^{-}` (lower case lambda) for each of these char speeds, instead of introducing a new symbol (tilde) to denote the same idea.

fmahebert commented on src/Evolution/Systems/ScalarWave/BoundaryCorrections/UpwindPenalty.hpp:
```diff
@@ -0,0 +1,288 @@
+ *    \left\{
+ *        \begin{array}{ll}
+ *          \lambda_{\hat{\alpha}} &
+ *            \mathrm{if}\;\lambda_{\hat{\alpha}}\le 0.0 \\
```
`0.0` -> `0` in a math equation

fmahebert commented on src/Evolution/Systems/ScalarWave/BoundaryCorrections/UpwindPenalty.hpp:
```diff
@@ -0,0 +1,288 @@
+ * \f}
+ *
+ * In the implementation we store the speeds in a rank-3 tensor with the zeroth
+ * component being \f$\lambda_{v^\Psi}\f$, the first is \f$\lambda_{v^+}\f$ and
```
"first is" -> "first being" to match other clauses in same sentence

fmahebert commented on src/Evolution/Systems/ScalarWave/BoundaryCorrections/UpwindPenalty.hpp:
```diff
@@ -0,0 +1,288 @@
+ *    \left\{
+ *        \begin{array}{ll}
+ *          -\lambda_{\hat{\alpha}} &
+ *            \mathrm{if}\;-\lambda_{\hat{\alpha}}\le 0.0 \\
```
Again "0.0"

fmahebert reviewed (commented):
My first "review" wasn't a full review, but was just my seeking to better understand a small point... so here's my review.

nilsdeppe commented:
I have pushed two more fixups. I did the `lambda` notation change in a separate fixup in case after seeing it you don't like it :)

fmahebert commented:
LGTM! I'm happy with the `tilde{lambda}` -> `lambda^-` change, but perhaps we should let @moxcodes have a say before you squash, since it is a non-trivial change?

moxcodes commented:
With the +/- superscripts, you now have the option of unifying the lambda definitions by using `\pm` and `\mp` as appropriate. I'll leave that up to you; either way I'm happy with the notation change and you can go ahead and squash.

nilsdeppe commented:
I've combined the definitions, but since it's the end of the day please double check I have the signs right :P I also:
- rebased
- added a call to register the boundary condition so the tests pass with the new serialization checks added there

moxcodes reviewed (approved):
Looks good -- at least I think the equations are consistent with their previous form.
